ffi/src/handle.rs

```diff
@@ -1,235 +1,426 @@
+    use std::sync::Arc;
+
+    use super::*;
+
+    /// Represents an object that crosses the FFI boundary and which outlives the scope that created
+    /// it. It remains valid until explicitly dropped by a call to [Handle::drop_handle].
     ///
-    /// The `handle` was returned by a call to `into_handle`, has not already been passed to
-    /// `drop_handle`, and has no other live references.
-    unsafe fn drop_handle(handle: *mut Self) {
-        let _ = unsafe { Box::from_raw(handle) };
+    /// cbindgen:transparent-typedef
+    #[repr(transparent)]
+    pub struct Handle<H: HandleDescriptor> {
+        ptr: NonNull<H>,
+    }
+
+    unsafe impl<H: HandleDescriptor> Send for Handle<H> {}
+    unsafe impl<H: HandleDescriptor> Sync for Handle<H> {}
+
+    impl<T, M, S, H> Handle<H>
+    where
+        T: ?Sized,
+        H: HandleDescriptor<Target = T, Mutable = M, Sized = S> + HandleOps<T, M, S>,
+    {
+        /// Obtains a shared reference to this handle's underlying object.
+        pub unsafe fn as_ref(&self) -> &H::Target {
+            H::as_ref(&self.ptr)
+        }
+
+        pub unsafe fn into_inner(self) -> H::From {
+            H::into_inner(self.ptr)
+        }
+        /// Destroys this handle.
+        pub unsafe fn drop_handle(self) {
+            drop(self.into_inner())
+        }
+    }
+
+    impl<T, S, H> From<Box<T>> for Handle<H>
+    where
+        T: ?Sized + Send + Sync,
+        H: HandleDescriptor<Target = T, Mutable = True, Sized = S>
+            + HandleOps<T, True, S, From = Box<T>>,
+    {
+        fn from(val: Box<T>) -> Handle<H> {
+            let ptr = H::into_handle(val).cast();
+            Handle { ptr }
+        }
+    }
+
+    impl<T, S, H> From<Arc<T>> for Handle<H>
+    where
+        T: ?Sized + Send + Sync,
+        H: HandleDescriptor<Target = T, Mutable = False, Sized = S>
+            + HandleOps<T, False, S, From = Arc<T>>,
+    {
+        fn from(val: Arc<T>) -> Handle<H> {
+            let ptr = H::into_handle(val).cast();
+            Handle { ptr }
+        }
+    }
+
+    impl<T, S, H> HandleAsMut for Handle<H>
+    where
+        T: ?Sized + Send + Sync,
+        H: HandleDescriptor<Target = T, Mutable = True, Sized = S> + HandleOps<T, True, S>,
+    {
+        type Target = T;
+
+        unsafe fn as_mut(&mut self) -> &mut T {
+            H::as_mut(&mut self.ptr)
+        }
+    }
+
+    impl<T, S, H> HandleAsArc for Handle<H>
+    where
+        T: ?Sized + Send + Sync,
+        H: HandleDescriptor<Target = T, Mutable = False, Sized = S> + HandleOps<T, False, S>,
+    {
+        type Target = T;
+
+        unsafe fn as_arc(&self) -> Arc<T> {
+            H::clone_arc(self.ptr)
+        }
+    }
+
+    pub trait HandleOps<T: ?Sized, M, S> {
+        type From: Sized;
+        type Raw: Sized;
+
+        fn into_handle(val: Self::From) -> NonNull<Self::Raw>;
+
+        unsafe fn as_ref(ptr: &NonNull<Self>) -> &T;
+
+        // WARNING: unimplemented when M=False
```

Comment:
Not quite sure what you mean? One of the two methods is already `unimplemented!()` in each of the specific impl for `H`. 

I intentionally didn't define trait-provided methods here, so that the compiler complains if any impl misses one.

Reply:
(found a way to avoid `unimplemented!` altogether)